sympy/polys/numberfields/subfield.py

```diff
@@ -0,0 +1,486 @@
+r"""
+Solve the "Subfield Problem" and allied problems, for algebraic number fields.
+
+Following Cohen (see [Cohen93] Section 4.5), we can define the problem as
+follows:
+
+* **Subfield Problem:**
+
+  Given two number fields $\mathbb{Q}(\alpha)$, $\mathbb{Q}(\beta)$
+  via the minimal polynomials for their generators $\alpha$ and $\beta$, decide
+  whether one field is isomorphic to a subfield of the other.
+
+From a solution to this problem flow solutions to the following problems as
+well:
+
+* **Primitive Element Problem:**
+
+  Given several algebraic numbers
+  $\alpha_1, \ldots, \alpha_m$, compute a single algebraic number $\theta$
+  such that $\mathbb{Q}(\alpha_1, \ldots, \alpha_m) = \mathbb{Q}(\theta)$.
+
+* **Field Isomorphism Problem:**
+
+  Decide whether two number fields
+  $\mathbb{Q}(\alpha)$, $\mathbb{Q}(\beta)$ are isomorphic.
+
+* **Field Membership Problem:**
+
+  Given two algebraic numbers $\alpha$,
+  $\beta$, decide whether $\alpha \in \mathbb{Q}(\beta)$, and if so write
+  $\alpha = f(\beta)$ for some $f(x) \in \mathbb{Q}[x]$.
+
+This module provides solutions to all these problems.
+"""
+
+from sympy.core.add import Add
+from sympy.core.numbers import AlgebraicNumber
+from sympy.core.singleton import S
+from sympy.core.symbol import Dummy
+from sympy.core.sympify import sympify
+from sympy.ntheory import sieve
+from sympy.polys.densetools import dup_eval
+from sympy.polys.domains import QQ
+from sympy.polys.numberfields.minpoly import _choose_factor, minimal_polynomial
+from sympy.polys.polyerrors import IsomorphismFailed
+from sympy.polys.polytools import Poly, PurePoly, factor_list
+from sympy.utilities import public
+
+from mpmath import pslq, mp
+
+
+def is_isomorphism_possible(a, b):
+    """Returns `True` if there is a chance for isomorphism. """
+    n = a.minpoly.degree()
+    m = b.minpoly.degree()
+
+    if m % n != 0:
+        return False
+
+    if n == m:
+        return True
+
+    da = a.minpoly.discriminant()
+    db = b.minpoly.discriminant()
+
+    i, k, half = 1, m//n, db//2
+
+    while True:
+        p = sieve[i]
+        P = p**k
+
+        if P > half:
+            break
+
+        if ((da % p) % 2) and not (db % P):
+            return False
+
+        i += 1
+
+    return True
+
+
+def field_isomorphism_pslq(a, b):
+    """Construct field isomorphism using PSLQ algorithm. """
+    if not a.root.is_real or not b.root.is_real:
+        raise NotImplementedError("PSLQ doesn't support complex coefficients")
+
+    f = a.minpoly
+    g = b.minpoly.replace(f.gen)
+
+    n, m, prev = 100, b.minpoly.degree(), None
+
+    for i in range(1, 5):
+        A = a.root.evalf(n)
+        B = b.root.evalf(n)
+
+        basis = [1, B] + [ B**i for i in range(2, m) ] + [A]
+
+        dps, mp.dps = mp.dps, n
+        coeffs = pslq(basis, maxcoeff=int(1e10), maxsteps=1000)
+        mp.dps = dps
```

Comment:
Does this change the global `dps` in mpmath? If so there should be a better way. At least this could use `try/finally` or a context manager but also is `dps` in mpmath threadsafe?

Is there an mpmath function that can be called that does not require altering the global dps?

Reply:
Actually I see now that you are just moving this function from one place to another. Still this isn't good.

Reply:
There are a few implementation details I wanted to revise in these modules as well, so maybe we can return to this and other points in another PR.